### PR TITLE
1332 install order

### PR DESCRIPTION
**Pull request by TrevCraw**
Fixes https://github.com/OpenLiberty/ci.maven/issues/1332

### CODE REVIEW

File: src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java [hunk 1]
```diff
@@ -1789,6 +1789,7 @@ public void restartServer(boolean buildContainer) throws PluginExecutionExceptio
         if (generateFeatures) {
             //TODO: decide if we should do an optimized feature generation on a server restart
             // If we do not generate here, need to revisit compileDependenciesChanged section in DevMojo
+            // If we generate here, we need to skip install features if there is a failure
             optimizeGenerateFeatures();
```
Comment (kathrynkodama) [inline src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java:1793]: Do we need to skip install features if there is a failure generating features? There may still be user specified features that we would want to install.

File: src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java [hunk 2]
```diff
@@ -3995,14 +4000,17 @@ private void processFileChanges(
...
+                boolean generateFeaturesSuccess = false;
                 if (generateFeatures) {
                     // custom server.xml modified
-                    incrementGenerateFeatures();
+                    generateFeaturesSuccess = incrementGenerateFeatures();
                 }
+                // suppress install feature warning - property must be set before calling copyConfigFolder
+                System.setProperty(SKIP_BETA_INSTALL_WARNING, Boolean.TRUE.toString());
+                // copyConfigFolder will install features if new ones are detected
+                copyConfigFolder(fileChanged, serverXmlFileParent, "server.xml", generateFeaturesSuccess);
```
Comment (kathrynkodama) [inline src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java:4011]: Even if generate features fails, dev mode should still install user specified features and copy their configuration file to target.

File: src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java [hunk 3]
```diff
@@ -4335,7 +4347,10 @@ public boolean accept(File pathname) {
             }
         }, true);
         copyFile(fileChanged, srcDir, tempConfig, targetFileName);
-        checkConfigFile(fileChanged, tempConfig);
+        // if feature generation is on, then it should succeed before installing features
+        if (!generateFeatures || generateFeaturesSuccess) {
+            checkConfigFile(fileChanged, tempConfig);
+        }
```
Comment (kathrynkodama) [inline src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java:4353]: May not be relevant given comments above, but it seems redundant to create the temporary configuration folder if we may be skipping `checkConfigFile`

File: src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java [hunk 4]
```diff
@@ -4029,20 +4038,22 @@ private void processFileChanges(
...
+                boolean generateFeaturesSuccess = false;
+                if ((fileChanged.getName().equals("server.xml")) && serverXmlFileParent == null && generateFeatures) {
+                    // server.xml modified
+                    generateFeaturesSuccess = incrementGenerateFeatures();
+                }
                 // suppress install feature warning - property must be set before calling copyConfigFolder
                 System.setProperty(SKIP_BETA_INSTALL_WARNING, Boolean.TRUE.toString());
-                copyConfigFolder(fileChanged, configDirectory, null);
+                // copyConfigFolder will install features if new ones are detected
+                copyConfigFolder(fileChanged, configDirectory, null, generateFeaturesSuccess);
```
Comment (kathrynkodama) [inline src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java:4050]: same as above, even if generate features fail we would still want to attempt to install any user specified features

File: src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java (hunk 1 above)
Comment (TrevCraw) [reply src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java:1793]: The expectation is that if we fail to generate features, it is due to conflicts somewhere and install feature will break anyway. We could investigate this further for the restart scenario (we are currently not skipping install).

File: src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java (hunk 2 above)
Comment (TrevCraw) [reply src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java:4011]: So far, in my testing, if generate features fails, then install features fails. It seems better to leave generate features as the error message for users to deal with in this situation. The current design in the UFO is that we do not install features after a failure to generate features. We can revisit this design if necessary.

Comment (kathrynkodama) [reply src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java:4011]: Sounds good, thank you

**Review by kathrynkodama — APPROVED** [review]

File: src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java (hunk 3 above)
Comment (TrevCraw) [reply src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java:4357]: Added TODO to revisit this in the future based on the decision for when to skip install features.

**Pull request merged**